fv3core/stencils/neg_adj3.py

```diff
@@ -1,184 +1,157 @@
-        dq = posqg if posqg < -qrtmp else -qrtmp
-        qrtmp = qrtmp + dq
-        qg = qg - dq
+        qliquid = 0.0
+        qrain_tmp = qsum
+        dq = pos_qgraupel if pos_qgraupel < -qrain_tmp else -qrain_tmp
+        qrain_tmp = qrain_tmp + dq
+        qgraupel = qgraupel - dq
         pt = pt - dq * icpk
-        # fill negative rain with available qi & qs (cooling)
-        if qr < 0.0:
-            dq = qi + qs if (qi + qs) < -qrtmp else -qrtmp
-            qrtmp = qrtmp + dq
-            dq1 = dq if dq < qs else qs
-            qs = qs - dq1
-            qi = qi + dq1 - dq
+        # fill negative rain with available qice & qsnow (cooling)
+        if qrain < 0.0:
+            dq = qice + qsnow if (qice + qsnow) < -qrain_tmp else -qrain_tmp
+            qrain_tmp = qrain_tmp + dq
+            dq1 = dq if dq < qsnow else qsnow
+            qsnow = qsnow - dq1
+            qice = qice + dq1 - dq
             pt = pt - dq * icpk
-        qr = qrtmp
+        qrain = qrain_tmp
         # fix negative rain water with available vapor
-        if qr < 0.0 and qv > 0.0:
-            dq = 0.999 * qv if 0.999 * qv < -qr else -qr
-            qv = qv - dq
-            qr = qr + dq
+        if qrain < 0.0 and qvapor > 0.0:
+            dq = 0.999 * qvapor if 0.999 * qvapor < -qrain else -qrain
+            qvapor = qvapor - dq
+            qrain = qrain + dq
             pt = pt + dq * lcpk
-    return qv, qi, qs, qg, qr, ql, pt
+    return qvapor, qice, qsnow, qgraupel, qrain, qliquid, pt
 
 
-# TODO, turn into stencil (started below, but need refactor to remove floats
-def fillq(q, dp, grid):
-    for j in range(grid.js, grid.je):
-        for i in range(grid.is_, grid.ie + 1):
-            sum1 = 0.0
-            for k in range(grid.npz):
-                if q[i, j, k] > 0.0:
-                    sum1 += q[i, j, k] * dp[i, j, k]
-            if sum1 < 1.0e-12:
-                continue
-            sum2 = 0.0
-            for k in range(grid.npz - 1, -1, -1):
-                if q[i, j, k] < 0.0 and sum1 > 0.0:
-                    dq = min(sum1, -q[i, j, k] * dp[i, j, k])
-                    sum1 -= dq
-                    sum2 += dq
-                    q[i, j, k] += dq / dp[i, j, k]
-            for k in range(grid.npz - 1, -1, -1):
-                if q[i, j, k] > 0.0 and sum2 > 0.0:
-                    dq = min(sum2, q[i, j, k] * dp[i, j, k])
-                    sum2 -= dq
-                    q[i, j, k] -= dq / dp[i, j, k]
-
-
-"""
-# TODO fix this to do fillq with a stencil that validates
-# need sum1, sum2 to be an accumulating floats
 @gtstencil()
-def fillq(q:sd, dp:sd):
+def fillq(q: FloatField, dp: FloatField, sum1: FloatFieldIJ, sum2: FloatFieldIJ):
+    with computation(FORWARD), interval(...):
+        # reset accumulating fields
+        sum1 = 0.0
+        sum2 = 0.0
     with computation(FORWARD), interval(...):
         if q > 0:
-            sum1 = sum1[0, 0, -1] + q * dp
+            sum1 = sum1 + q * dp
```

Comment:
This looks a bit suspect. Can you double check that it was intended?

Reply:
Sum1 and Sum2 are 2D fields now, so yes this was intended.